### PR TITLE
Add Assert.Multiple to the framework

### PR DESCRIPTION
**Pull request by CharliePoole**
Fixes issue #391 

This takes care of the syntactic part of Assert.Multiple. Note that we still may want to make format changes to the XML before our next release, but that can be done in a separate PR.

The original implementation used a MultipleAssertBlock class that lived in the TestExecutionContext and actually handled the reporting of errors. It seemed like a more object-oriented approach but introduced complications when the blocks were nested. I gradually weakened that class until it was doing nothing but counting entries into and exits from the blocks and ended up eliminating it entirely. Each TestExecutionContext now simply has a MultipleAssertLevel, which tracks nesting of blocks. We record the individual assertion failures in every case but only throw a final exception if the level is zero.

Writing tests that depended on finding the name of a particular method in the stack trace pointed out the need to disable optimization of the nunit.testdata assembly at least for certain platforms. It has been done now for all platforms just in case.

I have re-opened issue #1743 to allow us to have second thoughts - if necessary - on the XML format. The code in this PR is set up so that the actual elements we put out can be modified fairly simply if we decide to make a change.

### CODE REVIEW

File: src/NUnitFramework/testdata/nunit.testdata-3.5.csproj [hunk 1]
```diff
@@ -28,7 +28,7 @@
   </PropertyGroup>
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
     <DebugType>pdbonly</DebugType>
-    <Optimize>true</Optimize>
+    <Optimize>false</Optimize>
```
Comment (rprouse) [inline src/NUnitFramework/testdata/nunit.testdata-3.5.csproj]: If the optimization issue was only happening in .NET 2.0, do we need to disable in all projects? Doesn't really matter with the nature of this project, just asking...

File: src/NUnitFramework/framework/Assert.cs [hunk 2]
```diff
@@ -270,16 +272,95 @@ public static void Contains(object expected, ICollection actual)
...
+
+                if (count > 0)
+                {
+                    var writer = new TextMessageWriter("Multiple Assert block had {0} failure(s).", count);
+                    writer.WriteLine();
+
+                    int counter = 0;
+                    foreach (var assertion in context.CurrentResult.AssertionResults)
+                    {
+                        writer.WriteLine(string.Format("  {0}) {1}", ++counter, assertion.Message));
```
Comment (rprouse) [inline src/NUnitFramework/framework/Assert.cs]: Sorry, I should have been more explicit. The indent is fine, I was thinking you need the opening parenthesis.

**Comment by CharliePoole** [comment]
@rprouse I felt that @remcomulder's comment opened up a number possibilities but it also changed some things that I thought were already decided. This PR is primarily about the syntax anyway, and we don't seem to be arguing about that.

As I mentioned in the description, the code is set up to be easy to change. There's one place in the code to change the new representation and one place to change the "legacy" representation. I'd rather keep moving ahead and do a new PR if we decide to change either representation before a release. I re-opened #1743 for that purpose and I'll comment further there.

**Comment by rprouse** [comment]
@CharliePoole sounds good,that is what I thought, just wanted to double check. Good to merge if you intended the format as is. Looking at it again, I think you probably did.

**Comment by CharliePoole** [comment]
I restarted the build, which failed due to an unrelated timing error that's been giving us trouble. I'm merging anyway since I only removed a comment line in the last commit.

File: src/NUnitFramework/framework/Assert.cs [hunk 3]
```diff
@@ -270,16 +272,95 @@ public static void Contains(object expected, ICollection actual)
...
+
+#if PORTABLE
+            try
+            {
+                // Throw to get stack trace for recording the assertion
+                throw new Exception();
+            }
+            catch (Exception ex)
+            {
+                stackTrace = ex.StackTrace;
```
Comment (jcansdale) [inline src/NUnitFramework/framework/Assert.cs]: I think this is recording way too much stack information. For the `MultipleAssertFailureDemo` method, it records:

```
   at System.Environment.GetStackTrace(Exception e, Boolean needFileInfo)
   at System.Environment.get_StackTrace()
   at NUnit.Framework.Assert.ReportFailure(ConstraintResult result, String message, Object[] args) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Assert.cs:line 348
   at NUnit.Framework.Assert.That[TActual](TActual actual, IResolveConstraint expression, String message, Object[] args) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Assert.That.cs:line 247
   at NUnit.Framework.Assertions.Tests.AssertMultipleTests.<>c.<MultipleAssertFailureDemo>b__3_0() in C:\Source\GitHub\nunit\src\NUnitFramework\tests\Assertions\AssertMultipleTests.cs:line 100
   at NUnit.Framework.Assert.Multiple(TestDelegate testDelegate) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Assert.cs:line 290
   at NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo() in C:\Source\GitHub\nunit\src\NUnitFramework\tests\Assertions\AssertMultipleTests.cs:line 97
   at System.RuntimeMethodHandle._InvokeMethodFast(Object target, Object[] arguments, SignatureStruct& sig, MethodAttributes methodAttributes, RuntimeTypeHandle typeOwner)
   at System.RuntimeMethodHandle.InvokeMethodFast(Object target, Object[] arguments, Signature sig, MethodAttributes methodAttributes, RuntimeTypeHandle typeOwner)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture, Boolean skipVisibilityChecks)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at NUnit.Framework.Internal.Reflect.InvokeMethod(MethodInfo method, Object fixture, Object[] args) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Reflect.cs:line 206
   at NUnit.Framework.Internal.MethodWrapper.Invoke(Object fixture, Object[] args) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\MethodWrapper.cs:line 192
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunNonAsyncTestMethod(TestExecutionContext context) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 103
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(TestExecutionContext context) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 78
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:line 60
   at NUnit.Framework.Internal.Commands.TestActionCommand.Execute(TestExecutionContext context) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Commands\TestActionCommand.cs:line 81
   at NUnit.Framework.Internal.Commands.SetUpTearDownCommand.Execute(TestExecutionContext context) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Commands\SetUpTearDownCommand.cs:line 66
   at NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItem.cs:line 59
   at NUnit.Framework.Internal.Execution.WorkItem.RunTest() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 360
   at NUnit.Framework.Internal.Execution.WorkItem.Execute() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 260
   at NUnit.Framework.Internal.Execution.ParallelWorkItemDispatcher.Execute(WorkItem work) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\ParallelWorkItemDispatcher.cs:line 155
   at NUnit.Framework.Internal.Execution.ParallelWorkItemDispatcher.Dispatch(WorkItem work) in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\ParallelWorkItemDispatcher.cs:line 145
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs:line 237
   at NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs:line 117
   at NUnit.Framework.Internal.Execution.WorkItem.RunTest() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 360
   at NUnit.Framework.Internal.Execution.WorkItem.Execute() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs:line 260
   at NUnit.Framework.Internal.Execution.TestWorker.TestWorkerThreadProc() in C:\Source\GitHub\nunit\src\NUnitFramework\framework\Internal\Execution\TestWorker.cs:line 111
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
   at System.Threading.ExecutionContext.runTryCode(Object userData)
   at System.Runtime.CompilerServices.RuntimeHelpers.ExecuteCodeWithGuaranteedCleanup(TryCode code, CleanupCode backoutCode, Object userData)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart()
```

Could we add some filtering? I.e. remove everything before the Assert method and after 'NUnit.Framework.Assert.Multiple'?

**Comment by CharliePoole** [comment]
@jcansdale That's because our StackFilter removes NUnit.Framework.Assert entries from the top of the stack, but they are hidden under the System.Environment calls. This was fixed in PR #1890.

Are you reviewing master commits after the fact? We have found that sometimes leads to confusion and comes a bit too late in the development cycle when we are working quickly.

**Comment by jcansdale** [comment]
Sorry about that! I was looking at a local version. I'll pull and take another look.

**Comment by CharliePoole** [comment]
I just now merged PR #1890.

**Pull request merged**